src/Appwrite/Platform/Tasks/schedule.php

```diff
@@ -1,231 +1,221 @@
- *    if the function matches the current minute it should be dispatched to the functions worker.
- *    Then another translation is made to the cron pattern if it is in the next 5-min window
- *    it is assigned again to the  $queue. .
- * 4. Second timer  runs every X min and updates the $functions (large) list.
- *    The query fetches only functions that [resourceUpdatedAt] attr changed from the
- *    last time the timer that was fired (X min)
- *    If the function was deleted it is unsets from the list ($functions) and the $queue.
- *    In the end of the timer the $queue is created again.
- *
- */
-$cli
-->task('schedule')
-->desc('Function scheduler task')
-->action(function () use ($register) {
-    Console::title('Scheduler V1');
-    Console::success(APP_NAME . ' Scheduler v1 has started');
-
-    $dbForConsole = getConsoleDB();
+use function Swoole\Coroutine\run;
+
+class Schedule extends Action
+{
+    public const FUNCTION_UPDATE_TIMER = 10; //seconds
+    public const FUNCTION_ENQUEUE_TIMER = 60; //seconds
+
+    public static function getName(): string
+    {
+        return 'schedule';
+    }
+
+    public function __construct()
+    {
+        $this
+            ->desc('Execute functions scheduled in Appwrite')
+            ->inject('dbForConsole')
+            ->inject('getProjectDB')
+            ->callback(fn (Database $dbForConsole, callable $getProjectDB) => $this->action($dbForConsole, $getProjectDB));
+    }
 
     /**
-     * @return  void
-     */
-    $createQueue = function () use (&$functions, &$queue): void {
-        $loadStart = \microtime(true);
+     * 1. Load all documents from 'schedules' collection to create local copy
+     * 2. Create timer that sync all changes from 'schedules' collection to local copy. Only reading changes thanks to 'resourceUpdatedAt' attribute
+     * 3. Create timer that prepares coroutines for soon-to-execute schedules. When it's ready, coroutime sleeps until exact time before sending request to worker.
+    */
+    public function action(Database $dbForConsole, callable $getProjectDB): void
+    {
+        Console::title('Scheduler V1');
+        Console::success(APP_NAME . ' Scheduler v1 has started');
+
         /**
-         * Creating smaller functions list containing 5-min timeframe.
+         * Extract only nessessary attributes to lower memory used.
+         *
+         * @var Document $schedule
+         * @return  array
          */
-        $timeFrame = DateTime::addSeconds(new \DateTime(), FUNCTION_ENQUEUE_TIMEFRAME);
-        foreach ($functions as $function) {
-            $cron = new CronExpression($function['schedule']);
-            $next = DateTime::format($cron->getNextRunDate());
+        $getSchedule = function (Document $schedule) use ($dbForConsole, $getProjectDB): array {
+            $project = $dbForConsole->getDocument('projects', $schedule->getAttribute('projectId'));
+
+            [ $database, $reclaim ] = $getProjectDB($project);
```

Comment:
didn't we discuss to reclaim the entire pool and test it during the benchmarks ? We want to avoid this kind of multiple return values from functions. One of the reasons for the introduction of pools

Reply:
If the queue size was 100, and there were 101 changes, something would break, if not fully freeze. Considering schedule container is non-scalable, we should try to do reclaim logic properly.

With that said, if you want to avoid doing this anyway, I can remove it.

Reply:
Updated.